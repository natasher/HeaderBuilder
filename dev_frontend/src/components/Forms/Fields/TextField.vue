<template>
  <div class = "row">

    <div class = "col col-left">
      <span class = "title">{{ fieldName }}</span>
      <span class = "desc" v-if="spread === '' || typeof spread === 'undefined'">{{ desc }}</span>
    </div>

    <div class = "col col-right">
      <div :class = "[fieldDivClass, spread]">

        <input type        = "text"
              :name        = "wpId"
              :placeholder = "pholder"
              v-model      = "textFieldValue">

      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'

export default {

  name: 'TextField',

  props: {
    /**
     * The name of the field.
     * Displayed on the left column.
     */
    fieldName: {
      type    : String,
      required: true,
    },
    /**
     * Name of the field used by WordPress internals.
     */
    wpId: {
      type   : String,
      require: true,
    },
    /**
     * Description of the field.
     * Displayed under `fieldName`
     * if `spread` does *not* equal narrow.
     */
    desc: {
      type    : String,
      required: false,
      default : '',
    },
    /**
     * Determinates type of text field.
     * Default value is **empty**.
     * Possible value **narrow**.
     * If `spread=narrow` **desc** is shown.
     */
    spread: {
      type    : String,
      required: false,
      default : '',
    },
    /**
     * Placeholder value
     */
    pholder: {
      type    : String,
      required: false,
      default : '',
    },
    /**
     * Prop determinates type of Margin field,
     * possible values: **""**, **row**, **grid**.
     */
    as: {
      type    : String,
      required: false,
      default : '',
    },
    /**
     * When `as` prop is set to row, `row` indicates which one.
     * Then margin values will be set to given row.
     */
    row: {
      type    : String,
      required: false,
      default : '',
    },
  },

  data: function() {
    return {
      /**
       * Class added if prop `spread='narrow'`
       */
      fieldDivClass: 'mfn-field mfnf-text',
    }
  },

  methods: {
    ...mapActions( 'fields', [ 'setModalFieldValue' ]),
  },

  computed: {
    ...mapGetters( 'fields', [ 'getCurrentFieldValue', 'getRowOptionValue' ]),

    textFieldValue: {
      get() {
        if ( this.as === '' || typeof this.as == 'undefined' ) {

          return this.getCurrentFieldValue( this.wpId )

        } else if ( this.as === 'row' ) {

          return this.getRowOptionValue( this.wpId, this.row )

        }
      },
      set(value) {
        this.setModalFieldValue({
          name : this.wpId,
          value: value,
          as   : this.as,
          row  : this.row,
        })
      },
    },
  },
}
</script>
