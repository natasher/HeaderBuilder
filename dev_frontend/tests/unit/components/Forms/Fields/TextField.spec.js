import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import TextField     from '../../../../../src/components/Forms/Fields/TextField.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

const itemVal = 'item form'
const rowVal  = 'row value'

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      fields: {
        namespaced: true,
        actions: {
          setModalFieldValue: jest.fn()
        },
        getters: {
          getCurrentFieldValue: jest.fn().mockReturnValue(
            jest.fn().mockReturnValue( itemVal )
          ),
          getRowOptionValue: jest.fn().mockReturnValue(
            jest.fn().mockReturnValue( rowVal )
          ),
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    propsData: {
      fieldName: 'Test Field',
      wpId     : 'testField',
    }
  }

  return shallowMount( TextField, merge( defaultMountingOptions, overrides ))
}

describe('TextField.vue', () => {

  describe('snapshots:', () => {

    test('should match snapshot when field is wide', () => {
      const wrapper = createWrapper({
        propsData: {
          desc: 'Lorem ipsum dolor sit amet consectetur.'
        }
      })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot when field is narrow', () => {
      const wrapper = createWrapper({
        propsData: {
          desc  : 'Lorem ipsum dolor sit amet consectetur.',
          spread: 'narrow',
        }
      })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot when placeholder is provided', () => {
      const wrapper = createWrapper({
        propsData: {
          pholder: 'Lorem, ipsum dolor.'
        }
      })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  test('should get value from item form state if `as` prop is empty', async () => {
    const innerGetter = jest.fn().mockReturnValue('item form')
    const getCurrentFieldValue = jest.fn().mockReturnValue( innerGetter )
    const getRowOptionValue = jest.fn().mockReturnValue(
      jest.fn().mockReturnValue('row value')
    )
    const store = createStore({
      modules: {
        fields: {
          getters: {
            getCurrentFieldValue,
            getRowOptionValue,
          }
        }
      }
    })
    const wrapper = createWrapper({ store, })
    await flushPromises()

    expect(innerGetter).toBeCalledWith( wrapper.vm.wpId )
    expect(getRowOptionValue).not.toBeCalled()
  })

  test.only('should get value from row state if `as` props is `row`', async () => {
    const innerGetter          = jest.fn().mockReturnValue('row form')
    const getCurrentFieldValue = jest.fn().mockReturnValue( jest.fn() )
    const getRowOptionValue    = jest.fn().mockReturnValue( innerGetter )
    const store = createStore({
      modules: {
        fields: {
          getters: {
            getCurrentFieldValue,
            getRowOptionValue,
          }
        }
      }
    })
    const wrapper = createWrapper({
      store,
      propsData: {
        as: 'row'
      }
    })
    await flushPromises()

    expect(innerGetter).toBeCalledWith( wrapper.vm.wpId, '' )
    expect(getCurrentFieldValue).not.toBeCalled()
  })

  test('should set value if filled (item form ver.)', async () => {
    const setModalFieldValue = jest.fn()
    const store = createStore({
      modules: {
        fields: {
          actions: {
            setModalFieldValue,
          }
        }
      }
    })
    const wrapper = createWrapper({ store, })

    const input = wrapper.find( `input[name = "${ wrapper.vm.wpId}"]` )
    input.trigger( 'input' )
    await flushPromises()

    expect(setModalFieldValue).toBeCalledWith(
      expect.anything(),
      expect.objectContaining({
        name : wrapper.vm.wpId,
        value: itemVal,
      }),
      void 0
    )
  })

  test('should set value if filled (row ver.)', async () => {
    const setModalFieldValue = jest.fn()
    const store = createStore({
      modules: {
        fields: {
          actions: {
            setModalFieldValue,
          }
        }
      }
    })
    const wrapper = createWrapper({
      store,
      propsData: {
        as : 'row',
        row: 'testRow'
      }
    })

    const input = wrapper.find( `input[name = "${ wrapper.vm.wpId}"]` )
    input.trigger( 'input' )
    await flushPromises()

    expect(setModalFieldValue).toBeCalledWith(
      expect.anything(),
      expect.objectContaining({
        name : wrapper.vm.wpId,
        value: rowVal,
        as   : 'row',
        row  : 'testRow'
      }),
      void 0
    )
  })

})