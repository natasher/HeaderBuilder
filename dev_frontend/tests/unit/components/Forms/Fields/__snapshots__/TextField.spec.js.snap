// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField.vue snapshots: should match snapshot when field is narrow 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
     
    <!---->
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-text narrow"
    >
      <input
        name="testField"
        placeholder=""
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`TextField.vue snapshots: should match snapshot when field is wide 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
     
    <span
      class="desc"
    >
      Lorem ipsum dolor sit amet consectetur.
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-text"
    >
      <input
        name="testField"
        placeholder=""
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`TextField.vue snapshots: should match snapshot when placeholder is provided 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
     
    <span
      class="desc"
    >
      
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-text"
    >
      <input
        name="testField"
        placeholder="Lorem, ipsum dolor."
        type="text"
      />
    </div>
  </div>
</div>
`;
